### PR TITLE
Add setFilter and FilterRule

### PR DESCRIPTION
**Pull request by christophehenry**

### CODE REVIEW

File: lib/API.js [hunk 1]
```diff
@@ -75,5 +75,6 @@ export default {
   ServerCapabilities: require('./models/ServerCapabilities'),
   VacationResponse: require('./models/VacationResponse'),
   TransportError: require('./errors/TransportError'),
-  JmapError: require('./errors/JmapError')
+  JmapError: require('./errors/JmapError'),
+  FilterRule: require('./models/FilterRule'),
```
Comment (mbaechler) [inline lib/API.js]: wouldn't it make sense to have alphabetic ordering ?

File: lib/client/Client.js [hunk 2]
```diff
@@ -664,6 +665,30 @@ export default class Client {
     return Date.now();
   }
 
+  /**
+   * Sets the singleton {@link FilterRule} instance for a given account.<br />
+   * This will send a `setFilter` request to the JMAP backend.
+   *
+   * @param filteringRules {FilterRule[]} The {@link FilterRule} instance to set
```
Comment (mbaechler) [inline lib/client/Client.js]: It's very ambiguous, the param name is a plural, the type is an array and the description is not plural anymore

File: lib/models/FilterRule.js [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+      action: this.filterAction._toJSONObject()
+    };
+  }
+
+  static fromJSONObject(jmap, object) {
+    throw new Error('Not implemented');
+  }
+}
+
+FilterRule.ID = 'singleton'; // http://jmap.io/spec.html#vacation-response
```
Comment (mbaechler) [inline lib/models/FilterRule.js]: I don't think you want to link to vacation-response

File: lib/client/Client.js [hunk 4]
```diff
@@ -664,6 +665,34 @@ export default class Client {
...
+   * @param [options.accountId=null] {String} The account ID to set the filtering rules for. If `null`, the primary account is used.
+   *
+   * @returns {Promise} A {@link Promise} that eventually resolves to nothing upon success.
+   */
+  setFilter(filterRules, options) {
+    let filterRulesAsJson = filterRules.map(rule => rule instanceof FilterRule ? rule.toJSONObject() : rule);
+
+    return this._jmapRequest('setFilter', {
+      accountId: options && options.accountId,
+      //ifInState: options && options.ifInState || null,
```
Comment (tuanlc) [inline lib/client/Client.js]: Remove this comment if it is not needed

File: lib/client/Client.js [hunk 5]
```diff
@@ -664,6 +665,34 @@ export default class Client {
...
+   * This will send a `setFilter` request to the JMAP backend.
+   *
+   * @param filterRules {FilterRule[]|Object[]} The list of filter rules to set
+   * @param [options] {Object} The options to the underlying `setFilter` JMAP request.
+   * @param [options.accountId=null] {String} The account ID to set the filtering rules for. If `null`, the primary account is used.
+   *
+   * @returns {Promise} A {@link Promise} that eventually resolves to nothing upon success.
+   */
+  setFilter(filterRules, options) {
+    let filterRulesAsJson = filterRules.map(rule => rule instanceof FilterRule ? rule.toJSONObject() : rule);
```
Comment (tuanlc) [inline lib/client/Client.js]: Should be `const`

File: lib/utils/Utils.js [hunk 6]
```diff
@@ -144,7 +146,7 @@ export default class Utils {
    */
   static assertValidJMAPResponse(request, data) {
     function allArrayElementsAreArray(array) {
-      return array.filter(function(element) {
+      return array.filter(function (element) {
```
Comment (tuanlc) [inline lib/utils/Utils.js]: Why?

File: lib/utils/Utils.js [hunk 7]
```diff
@@ -203,7 +205,7 @@ export default class Utils {
       return uri;
     }
 
-    return uri.replace(/{(.+?)}/g, function(match, variable) {
+    return uri.replace(/{(.+?)}/g, function (match, variable) {
```
Comment (tuanlc) [inline lib/utils/Utils.js]: revert this please

File: lib/client/Client.js (hunk 4 above)
Comment (christophehenry) [reply lib/client/Client.js]: It is needed. Just lacking in current James' implementation.

Comment (tuanlc) [reply lib/client/Client.js]: Ok.

Comment (rouazana) [reply lib/client/Client.js]: it has been fixed on james side :)

**Pull request merged**